lib/main.dart

```diff
@@ -50,59 +50,145 @@ class App extends StatelessWidget {
-                    create: (context) {
-                      // Initialize T&C by loading the currently accepted version from shared preferences.
-                      final prefs = context.read<ServiceRepository>().sharedPreferences;
-                      return TermsAndConditionAcceptance(prefs);
-                    },
-                  ),
-                ],
-                child: MaterialApp(
-                  routes: appRoutes,
-                  theme: concordiumTheme(),
-                ),
-              );
-            },
-          ),
+          child: widget.child,
+        );
+      },
+    );
+  }
+}
+
+class _WithSelectedNetwork extends StatefulWidget {
+  final NetworkName initialNetwork;
+  final Widget child;
+
+  const _WithSelectedNetwork({required this.initialNetwork, required this.child});
+
+  @override
+  State<_WithSelectedNetwork> createState() => _WithSelectedNetworkState();
+}
+
+class _WithSelectedNetworkState extends State<_WithSelectedNetwork> {
+  late final Future<NetworkServices> _activating;
+
+  @override
+  void initState() {
+    super.initState();
+    final services = context.read<ServiceRepository>();
+    setState(() {
+      _activating = services.activateNetwork(initialNetwork);
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      future: _activating,
+      builder: (_, snapshot) {
+        final networkServices = snapshot.data;
+        if (networkServices == null) {
+          // Initializing network services.
+          return const _Initializing();
+        }
+
+        // Initialize blocs/cubits.
+        return BlocProvider(
+          create: (_) {
+            // Initialize selected network as the one that was just activated.
+            return SelectedNetwork(networkServices);
+          },
+          child: widget.child,
         );
       },
     );
   }
 }
 
+class _WithTermsAndConditionAcceptance extends StatefulWidget {
+  final Widget child;
+
+  const _WithTermsAndConditionAcceptance({required this.child});
+
+  @override
+  State<_WithTermsAndConditionAcceptance> createState() => _WithTermsAndConditionAcceptanceState();
+}
+
+class _WithTermsAndConditionAcceptanceState extends State<_WithTermsAndConditionAcceptance> {
+  late final Future<AcceptedTermsAndConditionsState?> _lastAccepted;
+  late final TermsAndConditionsRepository _repository;
+
+  @override
+  void initState() {
+    super.initState();
+    final storage = context.read<ServiceRepository>().storage;
+    setState(() {
+      _repository = TermsAndConditionsRepository(storageProvider: storage);
+      _lastAccepted = _repository.getAcceptedTermsAndConditions();
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+        future: _lastAccepted,
+        builder: (_, snapshot) {
+          // Check on the connection state since last accepted possible null
+          if (snapshot.connectionState != ConnectionState.done) {
```

Comment:
So to me this is a bit of a code smell when we check directly on the `connectionState`. According to one of Flutter's examples (https://docs.flutter.dev/cookbook/networking/fetch-data#5-display-the-data) the main methods to use are:
```
snapshot.hasData
snapshot.hasError
```
I understand that `hasData` might be false when the Future is resolved as it might return `null`. However, assuming that it completed successfully when `snapshot.connectionState == ConnectionState.done` will also lead to errors, as the state is also `done` when there is an error.

So the current implementation handles the error case and the successful case in the same way, and I think that may be hard to understand and misleading. I think it would be nicer to use a default value (avoid `null` values) and rely on `hasData` and `hasError`.

The same issue is present for `_WithSelectedNetworkState`.